### PR TITLE
[NeoMathEngine] blob convolution backward algorithms

### PR DESCRIPTION
**Pull request by favorart**
Algo1 -- Decrease RAM consumption and avoid excess matrix transposing -- speed-up
Algo2 -- Avoid excess matrix transposing -- speed-up

### CODE REVIEW

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 1]
```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
+	const int rrBufRowSize = source.Width() * filterObjectSize;
 
-	// Transpose the filter for easier calculations
-	// The filter depth represented by channels
-	const int filterForwardGeometricalSize = filter.Height() * filter.Width() * filter.Depth() * filter.Channels();
-	const int filterForwardChannelsCount = filter.BatchWidth();
-	const int filterForwardDataSize = filterForwardGeometricalSize * filterForwardChannelsCount;
-	CFloatHandleVar filterForward( mathEngine(), filterForwardDataSize );
-	float* filterForwardRaw = GetRaw( filterForward.GetHandle() );
+	constexpr int MaxThreadsCount = 4;
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Why 4? Even desktop processors may have 6 or 8 cores...

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 2]
```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
-		}
-		int filterRowBackStart = row - inputRowStart * desc.StrideHeight + desc.PaddingHeight;
-		if(0 > filterRowBackStart || filterRowBackStart >= filter.Height()) {
-			continue;
-		}
-		int filterRowBackEnd = filter.Height() + row - output.Height() - desc.PaddingHeight;
-		if(filterRowBackEnd < 0) {
-			filterRowBackEnd = 0;
-		}
+class CRoundRobinBuffer final {
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Need comment about class (what it does)

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 3]
```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
-				if( toCopy > 0 ) {
-					toCopy *= filterChannels;
-					tempRowDataShift *= filterChannels;
-					float* outputVec = outputDataPtr + pos * filterChannels;
-					vectorAdd( outputVec, tempRowData + tempRowDataShift, outputVec, toCopy );
-				}
-				tempRowData += filter.Height() * filter.Width() * filterChannels;
-			}
-		}
+	int GetExistItemsNum( int batchedSourceRowStart, int numItemsToAdd )
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: It's recommended to write multi-line methods outside of class declaration. The same goes for all multiline methods in this class.

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 4]
```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
-			continue;
-		}
-		int filterRowBackEnd = filter.Height() + row - output.Height() - desc.PaddingHeight;
-		if(filterRowBackEnd < 0) {
-			filterRowBackEnd = 0;
-		}
+class CRoundRobinBuffer final {
+public:
+	explicit CRoundRobinBuffer( IMathEngine& mathEngine, int numItems, int itemSize ) :
+		temp( mathEngine, /*bytes-size*/numItems * itemSize ), raw( NeoML::GetRaw( temp.GetHandle() ) ),
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: It's better to put each initializer on a separate line.

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 2 above)
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: You have a bunch of methods `Get*` without comments.

Some of them don't change the state of the object (`GetSourceRowNum`, `GetRaw` and `GetPrevBatchedSourceRowStart`).
Others are also `Get*` but change the state of the object (`GetExistItemsNum`, `GetBufPtr`)

At least add comments to each of them

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 5]
```diff
@@ -553,319 +536,337 @@ void CCpuMathEngine::BlobConvolution( const CConvolutionDesc& convDesc, const CC
...
+		return ptr;
+	}
+	const float* GetRaw( int offset ) const { return raw + offset; }
+	int GetPrevBatchedSourceRowStart() const { return prevBatchedSourceRowStart; }
+
+private:
+	CFloatHandleStackVar temp;
+	float* const raw;
+	const int numItems, itemSize;
+	int currPos /*start possition to add the buffer*/, prevBatchedSourceRowStart, prevNumItemsToAdd, startItem /*start of usefull data in buffer*/;
```
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Each field must be on a separate line.
And each field must be commented.

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 3 above)
Comment (FedyuninV) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: This method `Get`s something but `numItemsToAdd` says that something will be added... This class need _a lot_ of comments.

**Review by FedyuninV — CHANGES_REQUESTED** [review]

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 2 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 4 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 5 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 3 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: fixed

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 1 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Make value wider to 32

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp (hunk 2 above)
Comment (favorart) [reply NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Class has been renamed and refactored

**Review by FedyuninV — APPROVED** [review]

**Pull request merged**